src/scirpy/ir_dist/metrics.py

```diff
@@ -414,7 +363,362 @@ def _make_numba_matrix(distance_matrix: dict, alphabet: str = "ARNDCQEGHILKMFPST
     return dm
 
 
-class TCRdistDistanceCalculator:
+def _seqs2mat(
+    seqs: Sequence[str], alphabet: str = "ARNDCQEGHILKMFPSTWYVBZX", max_len: Union[None, int] = None
+) -> tuple[np.ndarray, np.ndarray]:
+    """Convert a collection of gene sequences into a
+    numpy matrix of integers for fast comparison.
+
+    Parameters
+    ----------
+    seqs:
+        Sequence of strings
+
+    Returns
+    -------
+    mat:
+        matrix with gene sequences encoded as integers
+    L:
+        vector with length values of the gene sequences in the matrix
+
+    Examples
+    --------
+    >>> seqs2mat(["CAT", "HAT"])
+    array([[ 4,  0, 16],
+        [ 8,  0, 16]], dtype=int8)
+
+    Notes
+    -----
+    Requires all seqs to have the same length, therefore shorter sequences
+    are filled up with -1 entries at the end.
+    """
+    if max_len is None:
+        max_len = np.max([len(s) for s in seqs])
+    mat = -1 * np.ones((len(seqs), max_len), dtype=np.int8)
+    L = np.zeros(len(seqs), dtype=np.int8)
+    for si, s in enumerate(seqs):
+        L[si] = len(s)
+        for aai in range(max_len):
+            if aai >= len(s):
+                break
+            try:
+                mat[si, aai] = alphabet.index(s[aai])
+            except ValueError:
+                # Unknown symbols given value for last column/row of matrix
+                mat[si, aai] = len(alphabet)
+    return mat, L
+
+
+class MetricDistanceCalculator(abc.ABC):
+    """
+    Abstract base class for distance calculator classes that compute parwise distances between
+    gene sequences in parallel based on a certain distance metric. The result is a (scipy) compressed sparse row distance matrix.
+    Derived classes just need to implement the method _metric_mat (see method comments for more details).
+
+    The code of this class is based on `pwseqdist <https://github.com/agartland/pwseqdist/blob/master/pwseqdist>`_.
+    Reused under MIT license, Copyright (c) 2020 Andrew Fiore-Gartland.
+
+    Parameters
+    ----------
+    n_jobs:
+        Number of threads per process to use for the pairwise distance calculation
+    n_blocks:
+        Overall number of blocks given to the workers (processes)
+    """
+
+    def __init__(self, n_jobs: int = 1, n_blocks: int = 1):
+        super().__init__()
+        self.n_jobs = n_jobs
```

Comment:
Done. nb.set_num_threads is now only used for positive integer values of self.n_jobs
```
if self.n_jobs > -1:
    nb.set_num_threads(self.n_jobs)
```
Here numba.config.NUMBA_NUM_THREADS is the default value if not explicitely set. Setting n_jobs to 0 or a higher value than available cores results in an error.

Reply:
One downside of using numba.config.NUMBA_NUM_THREADS however is that it might recognize all cores of a node of a computing cluster as available cores, even if your job on the cluster doesn't use all cores of the node (eg leo5).

Reply:
You are right. Please check my suggestion here if this makes sense to you: https://github.com/scverse/scirpy/pull/539/files

Then you could use `_get_usable_cpus(-1, use_numba=True)`.

Reply:
Good idea, I implemented it with this functions now!